src/controller/java/controller.i

```diff
@@ -733,6 +771,57 @@ namespace webots {
 %}
 %include <webots/PositionSensor.hpp>
 
+//----------------------------------------------------------------------------------------------
+//  Proto
+//----------------------------------------------------------------------------------------------
+
+%ignore webots::Proto::findProto(WbProtoRef ref);
+%ignore webots::Proto::cleanup();
+
+%rename("getParentPrivate") getParent() const;
+%javamethodmodifiers getParent() const "private"
+
+%rename("getFieldPrivate") getField(const std::string &fieldName) const;
+%rename("getFieldByIndexPrivate") getFieldByIndex(const int index) const;
+%javamethodmodifiers getField(const std::string &fieldName) const "private"
+%javamethodmodifiers getFieldByIndex(const int index) const "private"
+
+%typemap(javacode) webots::Proto %{
+// ----- begin hand written section ----
+  public Proto getParent() {
+    long cPtr = wrapperJNI.Proto_getParentPrivate(swigCPtr, this);
+    return Proto.findProto(cPtr);
+  }
+
+  public Field getField(String fieldName) {
+    long cPtr = wrapperJNI.Proto_getFieldPrivate(swigCPtr, this, fieldName);
+    return Field.findField(cPtr);
+  }
+
+  public Field getFieldByIndex(int index) {
+    long cPtr = wrapperJNI.Proto_getFieldByIndexPrivate(swigCPtr, this, index);
+    return Field.findField(cPtr);
+  }
+
+  private static java.util.HashMap<Long,Proto> protos = new java.util.HashMap<Long,Proto>();
+
+  // DO NOT USE THIS FUNCTION: IT IS RESERVED FOR INTERNAL USE !
+  public static Proto findProto(long cPtr) {
```

Comment:
Consider renaming to `findOrAddProto`.

Reply:
[See comment in `Proto.cpp`](https://github.com/cyberbotics/webots/pull/6613#discussion_r1747772547).